staging/src/k8s.io/apiserver/pkg/storage/errors.go

```diff
@@ -96,14 +99,40 @@ func NewInvalidObjError(key, msg string) *StorageError {
 	}
 }
 
+func NewCorruptedDataError(errToKey map[string]error) error {
+	return &StorageError{
+		Code:      ErrCodeCorruptedData,
+		KeyErrors: errToKey,
+	}
+}
+
 type StorageError struct {
 	Code               int
 	Key                string
+	KeyErrors          map[string]error
```

Comment:
An alternate that extends `StorageError` with an inner error:
```
--- a/staging/src/k8s.io/apiserver/pkg/storage/errors.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/errors.go
@@ -99,19 +99,20 @@ func NewInvalidObjError(key, msg string) *StorageError {
        }
 }
 
-func NewCorruptedDataError(errToKey map[string]error) error {
+func NewCorruptedDataError(key string, innerErr error) error {
        return &StorageError{
-               Code:      ErrCodeCorruptedData,
-               KeyErrors: errToKey,
+               Code:     ErrCodeCorruptedData,
+               Key:      key,
+               InnerErr: innerErr,
        }
 }
 
 type StorageError struct {
        Code               int
        Key                string
-       KeyErrors          map[string]error
        ResourceVersion    int64
        AdditionalErrorMsg string
+       InnerErr           error
 }
  
+func (e *StorageError) Unwrap() error { return e.InnerErr }
 ```

and inside etcd3/store.go, instead of using `failedKeys := make(map[string]error)`, we can use `errs := make([]error, 0)`, and then use aggregate error package https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go#L46.

Reply:
I like the idea that we hold on to the knowledge of keys for a little bit longer. I think we could implement unwrap to the aggregate error even with the current struct. Only I'm not sure if the aggregated error provides you with any more specific information, which I think is the usual reason why you would call it.